### PR TITLE
add query for filtering git app installations by software environment…

### PR DESCRIPTION
**Pull request by KPostOffice**
## Related Issues and Dependencies

thoth-station/thoth-application#1644

Create query to filter by software environment info. Adding issue creation to purge job is dependent on this change.

### CODE REVIEW

File: thoth/storages/graph/postgres.py [hunk 1]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+                query = query.filter(ExternalSoftwareEnvironment.image_name == image_name)
+            if image_sha:
+                query = query.filter(ExternalSoftwareEnvironment.image_sha == image_sha)
+            if os_name:
+                query = query.filter(ExternalSoftwareEnvironment.os_name == os_name)
+            if os_version:
+                os_version = normalize_os_version(os_name, os_version)
+                query = query.filter(ExternalSoftwareEnvironment.os_version == os_version)
+            if thoth_s2i_image_name:
+                query = query.filter(ExternalSofwareEnvironment.thoth_s2i_image_name == thoth_s2i_image_name)
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py:3667]: `ExternalSoftwareEnvironment` typo

File: thoth/storages/graph/postgres.py [hunk 2]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+            )
+
+            if python_version:
+                query = query.filter(ExternalSoftwareEnvironment.python_version == python_version)
+            if image_name:
+                query = query.filter(ExternalSoftwareEnvironment.image_name == image_name)
+            if image_sha:
+                query = query.filter(ExternalSoftwareEnvironment.image_sha == image_sha)
+            if os_name:
+                query = query.filter(ExternalSoftwareEnvironment.os_name == os_name)
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py:3662]: @fridex regarding https://github.com/thoth-station/storages/issues/2196#issuecomment-860744514, do we want to adjust the logic to have also only rhel and not ubi to be consistent? in that case here there should `map_os_name` method to have rhel if someone (bot or human) uses ubi8, which is what is present in .thoth.yaml and what purge job consider as source of truth in EcosystemSolver Table.

File: thoth/storages/graph/postgres.py [hunk 3]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+        image_sha: Optional[str] = None,
+        os_name: Optional[str] = None,
+        os_version: Optional[str] = None,
+        thoth_s2i_image_name: Optional[str] = None,
+        thoth_s2i_image_version: Optional[str] = None,
+        env_image_name: Optional[str] = None,
+        env_image_tag: Optional[str] = None,
+        cuda_version: Optional[str] = None,
+        environment_type: Optional[str] = None,
+    ):
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py:3647]: why no return?

File: thoth/storages/graph/postgres.py [hunk 4]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+            if thoth_s2i_image_version:
+                query = query.filter(ExternalSoftwareEnvironment.thoth_s2i_image_version == thoth_s2i_image_version)
+            if env_image_name:
+                query = query.filter(ExternalSoftwareEnvironment.env_image_name == env_image_name)
+            if env_image_tag:
+                query = query.filter(ExternalSoftwareEnvironment.cuda_version == cuda_version)
+            if environment_type:
+                query = query.filter(ExternalSoftwareEnvironment.environment_type == environment_type)
+
+            return query.all()
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py:3677]: Do you want to get objects here, not specific entities from the table?

File: thoth/storages/graph/postgres.py (hunk 2 above)
Comment (fridex) [reply thoth/storages/graph/postgres.py:3662]: +1 on the remapping

File: thoth/storages/graph/postgres.py [hunk 5]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
                 return active_managers
             return []
 
+    def get_kebechet_github_installation_info_for_software_environment(
```
Comment (fridex) [inline thoth/storages/graph/postgres.py:3633]: @pacospace should this be renamed to:
```suggestion
    def get_kebechet_github_installation_info_for_software_environment_all(
```

(to conform the naming convention)

File: thoth/storages/graph/postgres.py (hunk 4 above)
Comment (fridex) [inline thoth/storages/graph/postgres.py:3677]:
```suggestion
            return [i.to_dict() for i in query.all()]
```

File: thoth/storages/graph/postgres.py (hunk 2 above)
Comment (pacospace) [reply thoth/storages/graph/postgres.py:3662]: So we need to adjust sync logic in SoftwareEnvironment and ExternalSoftwareEnvironment Tables.. actually all tables that have `os_name` :D

File: thoth/storages/graph/postgres.py (hunk 5 above)
Comment (pacospace) [reply thoth/storages/graph/postgres.py:3633]: Yes exactly, Thanks @fridex !!

File: thoth/storages/graph/postgres.py (hunk 2 above)
Comment (fridex) [reply thoth/storages/graph/postgres.py:3662]: Sounds good 👍🏻

File: thoth/storages/graph/postgres.py [hunk 6]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
                 return active_managers
             return []
 
+    def get_kebechet_github_installation_info_for_software_environment_all(
```
[the remark this answers is not in the document]
Comment (KPostOffice) [reply thoth/storages/graph/postgres.py:3633]: Wouldn't this require adding filters for all possible values?

File: thoth/storages/graph/postgres.py [hunk 7]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+        env_image_name: Optional[str] = None,
+        env_image_tag: Optional[str] = None,
+        cuda_version: Optional[str] = None,
+        environment_type: Optional[str] = None,
+    ) -> List[Dict[str, Any]]:
+        with self._session_scope() as session, session.begin(subtransactions=True):
+            query = (
+                session.query(KebechetGithubAppInstallations)
+                .filter(KebechetGithubAppInstallations.is_active.is_(True))
+                .join(ExternalPythonSoftwareStack)
```
[the remark this answers is not in the document]
Comment (KPostOffice) [reply thoth/storages/graph/postgres.py:3652]: Sorry this is the wrong join. It should be `ExternalSoftwareEnvironment`

File: thoth/storages/graph/postgres.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pacospace) [reply thoth/storages/graph/postgres.py:3633]: not necessarily, we can add them when they are required :)

[the remark this answers is not in the document]
Comment (KPostOffice) [reply thoth/storages/graph/postgres.py]: Changed the function name as suggested

File: thoth/storages/graph/postgres.py [hunk 8]
```diff
@@ -3630,6 +3630,52 @@ def get_kebechet_github_installations_active_managers(self, slug: str) -> list:
...
+        os_name: Optional[str] = None,
+        os_version: Optional[str] = None,
+        thoth_s2i_image_name: Optional[str] = None,
+        thoth_s2i_image_version: Optional[str] = None,
+        env_image_name: Optional[str] = None,
+        env_image_tag: Optional[str] = None,
+        cuda_version: Optional[str] = None,
+        environment_type: Optional[str] = None,
+    ) -> List[Dict[str, Any]]:
+        with self._session_scope() as session, session.begin(subtransactions=True):
```
Comment (pacospace) [inline thoth/storages/graph/postgres.py:3648]: can you provide docstring with example please? similar to other methods:

```
"""Return info about repo containing Python Package in the software stack.
        Examples:
        >>> from thoth.storages import GraphDatabase
        >>> graph = GraphDatabase()
        >>> graph.get_kebechet_github_installations_info_for_python_package_version(
            package_name='click'
            index_url="https://pypi.org/simple",
        )
        {
            'thoth-station/jupyter-nbrequirements':
                {
                    'installation_id': '193650988',
                    'private': False,
                    'package_name': 'click',
                    'package_version': '7.1.2',
                    'index_url': 'https://pypi.org/simple'}
                }
        """
```

**Review by pacospace — APPROVED** [review]
Thanks 💯

**Pull request merged**